src/roaring64.c

```diff
@@ -0,0 +1,1366 @@
+    roaring64_bulk_context_t context = {0};
+    for (const uint64_t *current_val = vals; current_val != end;
+         current_val++) {
+        roaring64_bitmap_remove_bulk(r, &context, *current_val);
+    }
+}
+
+static inline void remove_range_closed_at(art_t *art, uint8_t *high48,
+                                          uint16_t min, uint16_t max) {
+    leaf_t *leaf = (leaf_t *)art_find(art, high48);
+    if (leaf == NULL) {
+        return;
+    }
+    uint8_t typecode2;
+    // container_add_range is exclusive but container_remove_range is
+    // inclusive...
+    container_t *container2 = container_remove_range(
+        leaf->container, leaf->typecode, min, max, &typecode2);
+    if (container2 != leaf->container) {
+        container_free(leaf->container, leaf->typecode);
+        leaf->container = container2;
+        leaf->typecode = typecode2;
+    }
+    if (!container_nonzero_cardinality(container2, typecode2)) {
+        art_erase(art, high48);
+        container_free(container2, typecode2);
+        free_leaf(leaf);
+    }
+}
+
+void roaring64_bitmap_remove_range_closed(roaring64_bitmap_t *r, uint64_t min,
+                                          uint64_t max) {
+    if (min > max) {
+        return;
+    }
+
+    art_t *art = &r->art;
+    uint8_t min_high48[ART_KEY_BYTES];
+    uint16_t min_low16 = split_key(min, min_high48);
+    uint8_t max_high48[ART_KEY_BYTES];
+    uint16_t max_low16 = split_key(max, max_high48);
+    if (compare_high48(min_high48, max_high48) == 0) {
+        // Only remove a range within one container.
+        remove_range_closed_at(art, min_high48, min_low16, max_low16);
+        return;
+    }
+
+    // Remove a range across containers. Remove intermediate containers
+    // entirely.
+    remove_range_closed_at(art, min_high48, min_low16, 0xffff);
+    uint64_t min_high_bits = min >> 16;
+    uint64_t max_high_bits = max >> 16;
+    for (uint64_t current = min_high_bits + 1; current < max_high_bits;
+         ++current) {
+        uint8_t current_high48[ART_KEY_BYTES];
+        split_key(current << 16, current_high48);
+        leaf_t *leaf = (leaf_t *)art_erase(art, current_high48);
+        if (leaf != NULL) {
+            container_free(leaf->container, leaf->typecode);
+            free_leaf(leaf);
+        }
+    }
+    remove_range_closed_at(art, max_high48, 0, max_low16);
+}
+
+uint64_t roaring64_bitmap_get_cardinality(const roaring64_bitmap_t *r) {
+    art_iterator_t it = art_init_iterator(&r->art, /*first=*/true);
+    uint64_t cardinality = 0;
+    while (it.value != NULL) {
+        leaf_t *leaf = (leaf_t *)it.value;
+        cardinality +=
+            container_get_cardinality(leaf->container, leaf->typecode);
+        art_iterator_next(&it);
+    }
+    return cardinality;
+}
+
+uint64_t roaring64_bitmap_range_cardinality(const roaring64_bitmap_t *r,
+                                            uint64_t min, uint64_t max) {
+    max--;  // A closed range is easier to work with.
```

Comment:
`roaring64_bitmap_range_cardinality(r, 0, 0)` should be valid, and return 0, this will instead compute the entire cardinality because of integer underflow